### ISSUE TITLE
PerRequestCacheManager Needs Proper Locking

### ISSUE THREAD

**Issue by toddca** [nopSolutions/nopCommerce#3529]
nopCommerce version: 4.10

Steps to reproduce the problem: 

I am using the REST API Plugin and I continually get the following errors/exceptions. The problem is due to the fact we also use Redis caching which uses PerRequestCaching and that class does not have proper thread protection - in my dev build I added reader/writer constructs and this fixed the issue.
```

Application: dotnet.exe
CoreCLR Version: 4.6.26919.2
Description: The process was terminated due to an unhandled exception.
Exception Info: System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
   at System.Collections.Generic.Dictionary`2.KeyCollection.Enumerator.MoveNext()
   at System.Linq.Enumerable.SelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.ToList()
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Nop.Core.Caching.PerRequestCacheManager.RemoveByPattern(String pattern) in 
   at Nop.Core.Caching.RedisCacheManager.RemoveByPatternAsync(String pattern) in 
   at Nop.Core.Caching.RedisCacheManager.RemoveByPattern(String pattern) in 
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location where exception was thrown ---
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
```

**Comment by skoshelev** [comment]
Closed #3529 (https://github.com/nopSolutions/nopCommerce/pull/3537)

**Comment by antoniodlp** [comment]
I am still seeing this issue in 4.40.4 using SQL Server distributed cache - how can we reopen it @skoshelev ?

**Comment by skoshelev** [comment]
Hi, @antoniodlp. Could you please provide us with more details about your problem? The full error stack would be a help to us.

**Comment by antoniodlp** [comment]
Hi @skoshelev, it is happening on a farm using SQL Server for distributed caching but I believe at the moment only two or one servers were on. I have not been able to reproduce it on my own. Here's the stack trace from the logs:

`System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
   at System.Collections.Generic.Dictionary`2.KeyCollection.Enumerator.MoveNext()
   at System.Linq.Enumerable.SelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.ToList()
   at Nop.Core.Caching.DistributedCacheManager.PerRequestCache.RemoveByPrefix(String prefix) in D:\a\1\s\Libraries\Nop.Core\Caching\DistributedCacheManager.cs:line 471
   at Nop.Core.Caching.DistributedCacheManager.RemoveByPrefixAsync(String prefix, Object[] prefixParameters) in D:\a\1\s\Libraries\Nop.Core\Caching\DistributedCacheManager.cs:line 283
   at Nop.Services.Caching.CacheEventConsumer`1.RemoveByPrefixAsync(String prefix, Object[] prefixParameters) in D:\a\1\s\Libraries\Nop.Services\Caching\CacheEventConsumer.cs:line 73
   at Nop.Services.Customers.Caching.CustomerCacheEventConsumer.ClearCacheAsync(Customer entity) in D:\a\1\s\Libraries\Nop.Services\Customers\Caching\CustomerCacheEventConsumer.cs:line 42
   at Nop.Services.Caching.CacheEventConsumer`1.ClearCacheAsync(TEntity entity, EntityEventType entityEventType) in D:\a\1\s\Libraries\Nop.Services\Caching\CacheEventConsumer.cs:line 52
   at Nop.Services.Caching.CacheEventConsumer`1.HandleEventAsync(EntityUpdatedEvent`1 eventMessage) in D:\a\1\s\Libraries\Nop.Services\Caching\CacheEventConsumer.cs:line 108
   at Nop.Services.Events.EventPublisher.PublishAsync[TEvent](TEvent event) in D:\a\1\s\Libraries\Nop.Services\Events\EventPublisher.cs:line 33`

**Comment by antoniodlp** [comment]
Hi @skoshelev, it is happening on a farm using SQL Server for distributed caching but I believe at the moment only two or one servers were on. I have not been able to reproduce it on my own. Here's the stack trace from the logs:

```
System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
   at System.Collections.Generic.Dictionary`2.KeyCollection.Enumerator.MoveNext()
   at System.Linq.Enumerable.SelectEnumerableIterator`2.MoveNext()
   at System.Linq.Enumerable.WhereEnumerableIterator`1.ToList()
   at Nop.Core.Caching.DistributedCacheManager.PerRequestCache.RemoveByPrefix(String prefix) in D:\a\1\s\Libraries\Nop.Core\Caching\DistributedCacheManager.cs:line 465
   at Nop.Core.Caching.DistributedCacheManager.RemoveByPrefixAsync(String prefix, Object[] prefixParameters) in D:\a\1\s\Libraries\Nop.Core\Caching\DistributedCacheManager.cs:line 277
   at Nop.Services.Caching.CacheEventConsumer`1.RemoveByPrefixAsync(String prefix, Object[] prefixParameters) in D:\a\1\s\Libraries\Nop.Services\Caching\CacheEventConsumer.cs:line 73
   at Nop.Services.Customers.Caching.CustomerCacheEventConsumer.ClearCacheAsync(Customer entity) in D:\a\1\s\Libraries\Nop.Services\Customers\Caching\CustomerCacheEventConsumer.cs:line 42
   at Nop.Services.Caching.CacheEventConsumer`1.ClearCacheAsync(TEntity entity, EntityEventType entityEventType) in D:\a\1\s\Libraries\Nop.Services\Caching\CacheEventConsumer.cs:line 52
   at Nop.Services.Caching.CacheEventConsumer`1.HandleEventAsync(EntityUpdatedEvent`1 eventMessage) in D:\a\1\s\Libraries\Nop.Services\Caching\CacheEventConsumer.cs:line 108
   at Nop.Services.Events.EventPublisher.PublishAsync[TEvent](TEvent event) in D:\a\1\s\Libraries\Nop.Services\Events\EventPublisher.cs:line 33
```

**Comment by skoshelev** [comment]
Closed #3529
